src/main/java/com/gluonhq/richtext/RichTextAreaSkin.java

```diff
@@ -152,23 +165,54 @@ public void dispose() {
         getSkinnable().setEditable(false); // removes all related listeners
         getSkinnable().editableProperty().removeListener(this::editableChangeListener);
         viewModel.removeChangeListener(textChangeListener);
+        backgroundColorPaths.removeListener(textBackgroundColorChangeListener);
     }
 
     /// PRIVATE METHODS /////////////////////////////////////////////////////////
 
-    //TODO Need more optimal way of rendering text fragments.
+    // TODO Need more optimal way of rendering text fragments.
     //  For now rebuilding the whole text flow
     private void refreshTextFlow() {
         fontCacheEvictionTimer.pause();
         try {
             var fragments = new ArrayList<Text>();
-            viewModel.walkFragments((text, decoration) -> fragments.add(buildText(text, decoration)));
+            var backgroundIndexRanges = new ArrayList<IndexRangeColor>();
+            var length = new AtomicInteger();
+            viewModel.walkFragments((text, decoration) -> {
+                final Text textNode = buildText(text, decoration);
+                fragments.add(textNode);
+
+                if (decoration.getBackground() != Color.TRANSPARENT) {
+                    final IndexRangeColor indexRangeColor = new IndexRangeColor(
+                            length.get(),
+                            length.get() + textNode.getText().length(),
+                            decoration.getBackground()
+                    );
+                    backgroundIndexRanges.add(indexRangeColor);
+                }
+                length.addAndGet(textNode.getText().length());
+            });
             textFlow.getChildren().setAll(fragments);
+            addBackgroundPathsToLayers(backgroundIndexRanges);
         } finally {
             fontCacheEvictionTimer.start();
         }
     }
 
+    private void addBackgroundPathsToLayers(List<IndexRangeColor> backgroundIndexRanges) {
+        final List<Path> backgroundPaths = backgroundIndexRanges.stream()
+                .map(indexRangeBackground -> {
+                    final Path path = new BackgroundColorPath(textFlow.rangeShape(indexRangeBackground.getStart(), indexRangeBackground.getEnd()));
```

Comment:
This still builds all the paths (nodes) on every refresh, and then requires layoutBounds to do the checks, which might be a costly operation.

The implementation looks good, however, and hopefully there won't be many background paths...

In any case, this is a side effect of not having the support built-in in JavaFX.

Reply:
Good point! I can actually listen to the changes made to the range and its respective color. It will remove any dependency on layoutBounds.

Incoming change.. ;)